tt_eager/tensor/tensor.cpp

```diff
@@ -44,6 +44,11 @@ Tensor::Tensor(const Storage storage, const ttnn::Shape shape, DataType dtype, L
                 tensor_impl::validate_on_device_dtype_and_layout(storage.buffer->device(), shape.value(), dtype, layout);
                 // Increment main thread ref count for all tensors on device
                 this->tensor_attributes->increment_main_thread_ref_count(this->workers.at(0));
+                // This tensor is being created from scratch in a worker. Track this and allow it to be explicitly
+                // deallocated inside the worker (composite ops do this).
+                if (not this->workers.at(0)->in_main_thread()) {
```

Comment:
Few questions as I am getting up to speed:

What is the relationship between tensor and worker? 
Same tensor can represent data sent to different tensix?
In what case worker is on a main thread - is it in case of the "legacy" mode?
How do we guarantee that all workers are on main or not on main thread?

Reply:
@tt-asaigal

Reply:
Hey @ayerofieiev-tt,

Each device has a worker thread bound to it. This manages host <--> device data movement and workload dispatch.

Wrt tensors on device, it made sense to make the device handle for these tensors a field we call `worker`. This is the tensor's handle to the worker thread, which makes it easier for us to asynchronously scheduler tensors `.to` `.cpu` and op calls on device.

Tensors can represent data interleaved/sharded across different tensix cores or interleaved across different DRAM banks.

Legacy mode, there are no workers. Everything is done by the main thread. If a tensor has workers specified, async mode will use those for specific operations. If not, the op/step will be performed by the main thread.

Hope this helps!